lib/elixir/lib/kernel/typespec.ex

```diff
@@ -587,10 +591,22 @@ defmodule Kernel.Typespec do
     end
   end
 
+  defp typespec({:record, _meta, _args}, _vars, caller, _state) do
+    compile_error(caller, "bad record type")
+  end
+
   # Handle ranges
-  defp typespec({:.., meta, args}, vars, caller, state) do
-    {args, state} = Enum.map_reduce(args, state, &typespec(&1, vars, caller, &2))
-    {{:type, line(meta), :range, args}, state}
+  defp typespec({:.., meta, [left, right]}, vars, caller, state) do
+    {left, state} = typespec(left, vars, caller, state)
+    {right, state} = typespec(right, vars, caller, state)
+
+    case {:erl_eval.partial_eval(left), :erl_eval.partial_eval(right)} do
+      {{:integer, _, first}, {:integer, _, last}} when first < last ->
```

Comment:
I included this because [erl_lint forces the "from/first" part of the range to be lower than the "to/last" part](https://github.com/erlang/otp/blob/master/lib/stdlib/src/erl_lint.erl#L2848).